pkg/faketikv/node.go

```diff
@@ -0,0 +1,212 @@
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
+// NodeState node's state.
+type NodeState int
+
+// some state
+const (
+	Up NodeState = iota
+	Down
+	LossConnect
+	Block
+)
+
+// Node simulates a TiKV.
+type Node struct {
+	*metapb.Store
+	sync.RWMutex
+	stats                   *pdpb.StoreStats
+	tick                    uint64
+	storeTick               int
+	regionTick              int
+	wg                      sync.WaitGroup
+	tasks                   map[uint64]Task
+	client                  Client
+	reciveRegionHeartbeatCh <-chan *pdpb.RegionHeartbeatResponse
+	ctx                     context.Context
+	cancel                  context.CancelFunc
+	state                   NodeState
+	// share cluster information
+	clusterInfo *ClusterInfo
+}
+
+// NewNode returns a Node.
+func NewNode(id uint64, addr string, pdAddr string) (*Node, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	store := &metapb.Store{
+		Id:      id,
+		Address: addr,
+	}
+	stats := &pdpb.StoreStats{
+		StoreId: id,
+		// TODO: configurable
+		Capacity:  1000000000000,
+		Available: 1000000000000,
+		StartTime: uint32(time.Now().Unix()),
+	}
+	client, reciveRegionHeartbeatCh, err := NewClient(pdAddr)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	return &Node{
+		Store:  store,
+		stats:  stats,
+		client: client,
+		ctx:    ctx,
+		cancel: cancel,
+		tasks:  make(map[uint64]Task),
+		state:  Down,
+		reciveRegionHeartbeatCh: reciveRegionHeartbeatCh,
+	}, nil
+}
+
+// Start starts the node.
+func (n *Node) Start() error {
+	ctx, cancel := context.WithTimeout(n.ctx, pdTimeout)
+	err := n.client.PutStore(ctx, n.Store)
+	cancel()
+	if err != nil {
+		return err
+	}
+	n.wg.Add(1)
+	go n.reciveRegionHeartbeat()
+	n.state = Up
+	return nil
+}
+
+func (n *Node) reciveRegionHeartbeat() {
+	defer n.wg.Done()
+	for {
+		select {
+		case resp := <-n.reciveRegionHeartbeatCh:
+			task := responseToTask(resp, n.clusterInfo)
+			if task != nil {
+				n.clusterInfo.AddTask(task)
+			}
+		case <-n.ctx.Done():
+			return
+		}
+	}
+}
+
+// Tick steps node status change.
+func (n *Node) Tick() {
+	if n.state != Up {
+		return
+	}
+	n.stepHeartBeat()
+	n.stepTask()
+	// step regions will step region status. like elect a leader, split region and so on.
+	n.clusterInfo.stepRegions()
+	n.tick++
+}
+
+// GetState returns current node state.
+func (n *Node) GetState() NodeState {
+	return n.state
+}
+
+func (n *Node) stepTask() {
+	n.Lock()
+	defer n.Unlock()
+	for _, task := range n.tasks {
+		task.Step(n.clusterInfo)
+		if task.IsFinished() {
+			log.Infof("[store %d][region %d] task finished: %s", n.GetId(), task.RegionID(), task.Desc())
+			n.clusterInfo.reportRegionChange(task.RegionID())
+			delete(n.tasks, task.RegionID())
+		}
+	}
+}
+
+func (n *Node) stepHeartBeat() {
+	n.storeTick = (n.storeTick + 1) % 10
+	n.regionTick = (n.regionTick + 1) % 60
```

Comment:
I think you can use _n.tick_ directly:
```
if n.tick%10 == 0 { ... }
if n.tick%60 == 0 { ... }
```
Also, define constants for 10 and 60, add configs if necessary.

Reply:
The rest LGTM.